subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheBuildServiceIntegrationTest.groovy

```diff
@@ -64,6 +67,109 @@ class ConfigurationCacheBuildServiceIntegrationTest extends AbstractConfiguratio
         outputDoesNotContain onFinishMessage
     }
 
+    def "build service is restored when using @ServiceReference"() {
+        given:
+        withCountingServicePlugin()
+        file('settings.gradle') << """
+            pluginManagement {
+                repositories {
+                    maven { url '$mavenRepo.uri' }
+                }
+            }
+        """
+        file('build.gradle') << """
+            plugins { id 'counting-service-plugin' version '1.0' }
+
+            tasks.register('count', CountingTask) {
+                doLast {
+                    assert countingService.get().increment() == 2
+                }
+            }
+        """
+        def configurationCache = newConfigurationCacheFixture()
+
+        when:
+        configurationCacheRun(":count")
+
+        then:
+        configurationCache.assertStateStored()
+        outputContains 'Count: 1'
+
+        when:
+        configurationCacheRun(":count")
+
+        then:
+        configurationCache.assertStateLoaded()
+        outputContains 'Count: 1'
+    }
+
+    private void withCountingServicePlugin() {
+        createDir('plugin') {
+            file("src/main/java/CountingServicePlugin.java") << """
+                public abstract class CountingServicePlugin implements $Plugin.name<$Project.name> {
+                    private $Provider.name<?> counterProvider;
+                    @Override
+                    public void apply($Project.name project) {
+                        project.getGradle().getSharedServices().registerIfAbsent(
+                            "counter",
+                            CountingService.class,
+                            (spec) -> {}
+                        );
+                    }
+                }
+
+                abstract class CountingTask extends $DefaultTask.name {
+
+                    @$ServiceReference.name("counter")
+                    public abstract $Property.name<CountingService> getCountingService();
+
+                    @$Inject.name
```

Comment:
❓ Is `@Inject` necessary here at all? Is a constructor even required?

Reply:
I had copied from the other services in this class assuming it was needed. You are right, it isn't. However, as I started removing from the other services, I saw this for the service used in the tests that exercises included builds vs. classloaders:

(I wonder if we only use JSR330 instantiation for that case)

```
> Task :included:classloader1:probe FAILED
Could not stop org.gradle.internal.build.event.DefaultBuildEventsListenerRegistry$ForwardingBuildEventConsumer@1c04cb14.
org.gradle.api.services.internal.ServiceLifecycleException: Failed to create service 'listener of classloader2'.
	at org.gradle.api.services.internal.RegisteredBuildServiceProvider.instantiationException(RegisteredBuildServiceProvider.java:125)
	at org.gradle.api.services.internal.RegisteredBuildServiceProvider.instantiate(RegisteredBuildServiceProvider.java:120)
	at org.gradle.api.services.internal.RegisteredBuildServiceProvider.getInstance(RegisteredBuildServiceProvider.java:105)
	at org.gradle.api.services.internal.RegisteredBuildServiceProvider.calculateOwnValue(RegisteredBuildServiceProvider.java:98)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateOwnPresentValue(AbstractMinimalProvider.java:73)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:93)
	at org.gradle.internal.build.event.DefaultBuildEventsListenerRegistry$ForwardingBuildEventConsumer.handle(DefaultBuildEventsListenerRegistry.java:251)
	at org.gradle.internal.build.event.DefaultBuildEventsListenerRegistry$ForwardingBuildEventConsumer.handle(DefaultBuildEventsListenerRegistry.java:228)
	at org.gradle.internal.build.event.DefaultBuildEventsListenerRegistry$AbstractListener.run(DefaultBuildEventsListenerRegistry.java:170)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.gradle.api.reflect.ObjectInstantiationException: Could not create an instance of type my.ProbeService.
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:69)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.newInstance(DependencyInjectingInstantiator.java:55)
	at org.gradle.api.services.internal.RegisteredBuildServiceProvider.instantiate(RegisteredBuildServiceProvider.java:129)
	at org.gradle.api.services.internal.RegisteredBuildServiceProvider.instantiate(RegisteredBuildServiceProvider.java:118)
	... 14 more
Caused by: java.lang.IllegalArgumentException: Class ProbeService has no constructor that is annotated with @Inject.
	at org.gradle.internal.instantiation.generator.InjectUtil.selectConstructor(InjectUtil.java:72)
	at org.gradle.internal.instantiation.generator.Jsr330ConstructorSelector.lambda$forType$0(Jsr330ConstructorSelector.java:57)
	at org.gradle.cache.Cache.lambda$get$0(Cache.java:31)
	at org.gradle.cache.internal.DefaultCrossBuildInMemoryCacheFactory$AbstractCrossBuildInMemoryCache.get(DefaultCrossBuildInMemoryCacheFactory.java:130)
	at org.gradle.cache.Cache.get(Cache.java:31)
	at org.gradle.internal.instantiation.generator.Jsr330ConstructorSelector.forType(Jsr330ConstructorSelector.java:53)
	at org.gradle.internal.instantiation.generator.Jsr330ConstructorSelector.forParams(Jsr330ConstructorSelector.java:48)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:61)
	... 17 more
```

Reply:
Never mind, please ignore the above. A constructor is always required, the `Inject` annotation is not.